compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala

```diff
@@ -0,0 +1,914 @@
+        sym.owner.companionModule
+      }
+      companionModule.moduleClass*/
+      sym.owner.companionModule.moduleClass
+    } else {
+      sym.owner
+    }
+  }
+
+  private def defaultGetterDenot(targetSym: Symbol, sym: Symbol, paramIndex: Int): Denotation =
+    targetSym.info.member(DefaultGetterName(sym.name.asTermName, paramIndex))
+
+  private def defaultGetterDenot(sym: Symbol, paramIndex: Int): Denotation =
+    defaultGetterDenot(targetSymForDefaultGetter(sym), sym, paramIndex)
+
+  /** Generate the final forwarding call to the exported method. */
+  private def genResult(exported: Exported, args: List[js.Tree], static: Boolean)(
+      implicit pos: SourcePosition): js.Tree = {
+
+    val sym = exported.sym
+
+    def receiver = {
+      if (static)
+        genLoadModule(sym.owner)
+      else if (sym.owner == defn.ObjectClass)
+        js.This()(jstpe.ClassType(ir.Names.ObjectClass))
+      else
+        js.This()(encodeClassType(sym.owner))
+    }
+
+    def boxIfNeeded(call: js.Tree): js.Tree = {
+      box(call, atPhase(elimErasedValueTypePhase)(sym.info.resultType))
+    }
+
+    if (currentClassSym.isNonNativeJSClass) {
+      assert(sym.owner == currentClassSym.get, sym.fullName)
+      boxIfNeeded(genApplyJSClassMethod(receiver, sym, args))
+    } else {
+      if (sym.isClassConstructor)
+        js.New(encodeClassName(currentClassSym), encodeMethodSym(sym), args)
+      else if (sym.isPrivate)
+        boxIfNeeded(genApplyMethodStatically(receiver, sym, args))
+      else
+        boxIfNeeded(genApplyMethod(receiver, sym, args))
+    }
+  }
+
+  private def genThrowTypeError(msg: String = "No matching overload")(implicit pos: Position): js.Tree =
+    js.Throw(js.JSNew(js.JSGlobalRef("TypeError"), js.StringLiteral(msg) :: Nil))
+
+  private final class ParamSpec(val sym: Symbol, val info: Type,
+      val isRepeated: Boolean, val hasDefault: Boolean) {
+    override def toString(): String =
+      s"ParamSpec(${sym.name}, ${info.show}, isRepeated = $isRepeated, hasDefault = $hasDefault)"
+  }
+
+  private object ParamSpec {
+    def apply(methodSym: Symbol, sym: Symbol, infoAtElimRepeated: Type, infoAtElimEVT: Type,
+        methodHasDefaultParams: Boolean, paramIndex: Int): ParamSpec = {
+      val isRepeated = infoAtElimRepeated.isRepeatedParam
+      val info = if (isRepeated) infoAtElimRepeated.repeatedToSingle else infoAtElimEVT
+      val hasDefault = methodHasDefaultParams && defaultGetterDenot(methodSym, paramIndex).exists
+      new ParamSpec(sym, info, isRepeated, hasDefault)
+    }
+  }
+
+  // This is a case class because we rely on its structural equality
+  private final case class Exported(sym: Symbol) {
+    private val isAnonJSClassConstructor =
+      //sym.isClassConstructor && sym.owner.isAnonymousClass && isJSType(sym.owner)
+      false
+
+    val isLiftedJSConstructor =
+      sym.isClassConstructor && sym.owner.isNestedJSClass
+
+    val (params, captureParamsFront, captureParamsBack) = {
```

Comment:
In scalac, all three of those are `List[ParamSpec]`. In dotc the last 2 are `List[js.ParamDef]`, because it seems we cannot reliably obtain `Symbol`s for parameters added by intermediate phases (in this case, erasure and lambdalift).

Somehow it looks like this computation has become even uglier than it was in scalac ...

Reply:
I have now simplified this a lot. It looks better than what we have in Scala 2 now :)